samples/msal-node-samples/nextjs-rsc-ts/src/utils/AuthProvider.ts

```diff
@@ -0,0 +1,174 @@
+import {
+  AuthorizationCodeRequest,
+  AuthorizationUrlRequest,
+  ConfidentialClientApplication,
+  Configuration,
+  CryptoProvider,
+  DistributedCachePlugin,
+  ICacheClient,
+  IPartitionManager,
+} from "@azure/msal-node";
+import { cache } from "react";
+
+export type PartitionManagerFactory = () => Promise<IPartitionManager>;
+
+type AuthCodeRequestState = {
+  returnTo: string;
+  request: AuthorizationCodeRequest;
+};
+
+/**
+ * Light wrapper for msal-node's ConfidentialClientApplication.
+ */
+export class AuthProvider {
+  configuration: Configuration;
+  cacheClient: ICacheClient;
+  partitionManagerFactory: PartitionManagerFactory;
+  cryptoProvider: CryptoProvider;
+  redirectUri: string;
+
+  /**
+   * Initialize the AuthProvider.
+   * @param configuration The msal configuration object
+   * @param redirectUri Uri that authentication requests will redirect back to
+   * @param cacheClient The cache client used to store the token cache
+   * @param partitionManagerFactory Factory that returns a PartitionManager for the current executing context
+   */
+  constructor(
+    configuration: Configuration,
+    redirectUri: string,
+    cacheClient: ICacheClient,
+    partitionManagerFactory: PartitionManagerFactory
+  ) {
+    this.configuration = configuration;
+    this.redirectUri = redirectUri;
+    this.cacheClient = cacheClient;
+    this.partitionManagerFactory = cache(partitionManagerFactory);
+
+    this.cryptoProvider = new CryptoProvider();
+  }
+
+  /**
+   * Get url for an auth code request
+   * @param request Authorization request to initialize the flow
+   * @param returnTo Where the user should be redirected to after a successful flow
+   * @returns The url to redirect the client to
+   */
+  async getAuthCodeUrl(
+    request: Omit<AuthorizationUrlRequest, "redirectUri">,
+    returnTo: string
+  ) {
+    const instance = await this.getInstance();
+
+    const requestWithRedirectUri = {
+      ...request,
+      redirectUri: this.redirectUri,
+    };
+
+    const state = this.cryptoProvider.base64Encode(
+      JSON.stringify({
+        request: requestWithRedirectUri,
+        returnTo,
+      })
+    );
+
+    return await instance.getAuthCodeUrl({
+      ...requestWithRedirectUri,
+      state: state,
+    });
+  }
+
+  /**
+   * Handles token acquisition based on the url that the user was sent to from Azure.
+   * @param url The return url from Azure
+   * @returns An object containing the logged in account, and where the user should be redirected to.
+   */
+  async handleAuthCodeCallback(url: URL) {
+    const instance = await this.getInstance();
```

Comment:
A new CCA instance will come with an empty metadata cache. See this for more: [performance-considerations-for-confidential-client-applications](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/dev/lib/msal-node/docs/performance.md#performance-considerations-for-confidential-client-applications). There's an ongoing discussion on what's the best way to cache metadata for web apps, so leaving this here as a reminder, but something [like this](https://github.com/AzureAD/microsoft-authentication-library-for-js/blob/a29f54835d039b109929bad313d488c29d7a1485/samples/msal-node-samples/auth-code-distributed-cache/src/AuthProvider.ts#L135) might be here.